### PR TITLE
[Merged by Bors] - Support async response in socket

### PR DESCRIPTION
**Pull request by galibey**
Added an ability to receive a response asynchronously. We re-use the current functionality of `AsyncResponse` and streaming, just saying that `AsyncResponse` is now also a Future that can resolve to one result instead of a sequence of results for the streaming case.

Related #2481

### CODE REVIEW

File: crates/fluvio-socket/src/multiplexing.rs [hunk 1]
```diff
@@ -459,27 +482,31 @@ impl MultiPlexingResponseDispatcher {
...
-                            format!(
-                                "problem sending to queue socket: {}, err: {}",
-                                correlation_id, err
-                            ),
-                        )
-                        .into()
-                    })
+                    // sender was dropped before response arrives
+                    if queue_sender.is_closed() {
+                        debug!("attempt to send data to closed socket: {}", correlation_id);
```
Comment (sehz) [inline crates/fluvio-socket/src/multiplexing.rs:487]: Can you use trace format? 
`debug!(correlation_id,"attemp....")`

File: crates/fluvio-socket/src/multiplexing.rs [hunk 2]
```diff
@@ -459,27 +482,31 @@ impl MultiPlexingResponseDispatcher {
...
-            Err(IoError::new(
-                ErrorKind::BrokenPipe,
-                format!(
-                    "no socket receiver founded for id: {}, abandoning sending",
-                    correlation_id
-                ),
-            )
-            .into())
+            // sender was dropped and unregistered before response arrives
+            debug!(
```
Comment (sehz) [inline crates/fluvio-socket/src/multiplexing.rs:505]: use trace lib format. see comment above

File: crates/fluvio-socket/Cargo.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "fluvio-socket"
-version = "0.12.0"
+version = "0.13.0"
```
Comment (sehz) [inline crates/fluvio-socket/Cargo.toml:3]: Is there API incompatibility? If we are just adding a new API, then just need to bump up patch version

**Review by sehz — CHANGES_REQUESTED** [review]
Overall looks good.  Just few minor nits

File: crates/fluvio-socket/src/multiplexing.rs (hunk 1 above)
Comment (galibey) [reply crates/fluvio-socket/src/multiplexing.rs]: Done

File: crates/fluvio-socket/src/multiplexing.rs (hunk 2 above)
Comment (galibey) [reply crates/fluvio-socket/src/multiplexing.rs:505]: Done

File: crates/fluvio-socket/Cargo.toml (hunk 3 above)
Comment (galibey) [reply crates/fluvio-socket/Cargo.toml]: I thought that implementing `Future`  on `AsyncResponse` potentially could be a breaking change, but `orphan rules` would not allow it, so you're right. Changed to 0.12.1.

**Review by sehz — APPROVED** [review]
LGTM

**Pull request closed without merging**